src/hotspot/share/gc/parallel/parallelScavengeHeap.cpp

```diff
@@ -427,74 +419,81 @@ HeapWord* ParallelScavengeHeap::mem_allocate_old_gen(size_t size) {
+}
 
-  // We assume that allocation in eden will fail unless we collect.
+void ParallelScavengeHeap::do_full_collection_no_gc_locker(bool clear_all_soft_refs) {
+  bool maximum_compaction = clear_all_soft_refs;
+  PSParallelCompact::invoke(maximum_compaction);
+}
 
-  // First level allocation failure, scavenge and allocate in young gen.
-  GCCauseSetter gccs(this, GCCause::_allocation_failure);
-  const bool invoked_full_gc = PSScavenge::invoke();
-  HeapWord* result = young_gen()->allocate(size);
+HeapWord* ParallelScavengeHeap::expand_heap_and_allocate(size_t size, bool is_tlab) {
+  HeapWord* result = nullptr;
 
-  // Second level allocation failure.
-  //   Mark sweep and allocate in young generation.
-  if (result == nullptr && !invoked_full_gc) {
-    do_full_collection(false);
-    result = young_gen()->allocate(size);
+  result = young_gen()->allocate(size);
+  if (result == nullptr && !is_tlab) {
+    // auto expand inside
+    result = old_gen()->allocate(size);
   }
+  return result;   // Could be null if we are out of space.
+}
+
+HeapWord* ParallelScavengeHeap::satisfy_failed_allocation(size_t size, bool is_tlab) {
+  assert(size != 0, "precondition");
 
-  death_march_check(result, size);
+  HeapWord* result = nullptr;
 
-  // Third level allocation failure.
-  //   After mark sweep and young generation allocation failure,
-  //   allocate in old generation.
-  if (result == nullptr) {
-    result = allocate_old_gen_and_record(size);
+  GCLocker::check_active_before_gc();
+  if (GCLocker::is_active_and_needs_gc()) {
+    return expand_heap_and_allocate(size, is_tlab);
   }
 
-  // Fourth level allocation failure. We're running out of memory.
-  //   More complete mark sweep and allocate in young generation.
-  if (result == nullptr) {
-    do_full_collection(true);
-    result = young_gen()->allocate(size);
+  // If young-gen can handle this allocation, attempt young-gc firstly.
+  bool should_run_young_gc = is_tlab || should_alloc_in_eden(size);
+  collect_at_safepoint(!should_run_young_gc);
+
+  result = expand_heap_and_allocate(size, is_tlab);
+  if (result != nullptr) {
+    return result;
   }
 
-  // Fifth level allocation failure.
-  //   After more complete mark sweep, allocate in old generation.
-  if (result == nullptr) {
-    result = allocate_old_gen_and_record(size);
+  // If we reach this point, we're really out of memory. Try every trick
+  // we can to reclaim memory. Force collection of soft references. Force
+  // a complete compaction of the heap. Any additional methods for finding
+  // free memory should be here, especially if they are expensive. If this
+  // attempt fails, an OOM exception will be thrown.
+  {
+    // Make sure the heap is fully compacted
+    uintx old_interval = HeapMaximumCompactionInterval;
+    HeapMaximumCompactionInterval = 0;
+
+    const bool clear_all_soft_refs = true;
+    do_full_collection_no_gc_locker(clear_all_soft_refs);
```

Comment:
If the young collection succeeded in method `collect_at_safepoint`. The normal full collection won't run in `collect_at_safepoint`. If the successful young collection didn't release any memory (or only released little memory but not enough for allocation), the allocation in line 462 will fail too. Then a full collection with maximum compaction will be run. It is strange. In my opinion, I think the steps look like below:

1. allocation
2. young collection
3. allocation
4. normal full collection
5. allocation
6. maximum full collection
7. allocation
8. OOM

But in current patch, the step 4-5 may be skipped.

Reply:
> If the successful young collection didn't release any memory (or only released little memory but not enough for allocation),

A successful young-gc often leave young-gen completely empty. Otherwise, max-compaction full-gc should be run -- there is little benefit of running non-max-compaction full-gc if old-gen is too packed to hold all young-gen objs.

Reply:
Thanks for your explanation. I am OK with the current solution now.